ignite/engine/engine.py

```diff
@@ -123,28 +123,42 @@ def compute_mean_std(engine, batch):
     _state_dict_one_of_opt_keys = ("iteration", "epoch")
 
     def __init__(self, process_function: Callable):
-        self._event_handlers = defaultdict(list)  # type: Dict[Any, List]
+        super(Engine, self).__init__()
         self.logger = logging.getLogger(__name__ + "." + self.__class__.__name__)
         self._process_function = process_function
-        self.last_event_name = None  # type: Optional[Events]
         self.should_terminate = False
         self.should_terminate_single_epoch = False
-        self.state = State()
+        self._state = State(engine=self)
         self._state_dict_user_keys = []  # type: List[str]
-        self._allowed_events = []  # type: List[EventEnum]
 
         self._dataloader_iter = None  # type: Optional[Iterator[Any]]
         self._init_iter = []  # type: List[int]
 
-        self.register_events(*Events)
+        self.register_events(*Events, attr_to_events=State.attr_to_events)
 
         if self._process_function is None:
             raise ValueError("Engine must be given a processing function in order to run.")
 
         _check_signature(process_function, "process_function", self, None)
 
+    @property
+    def state(self) -> State:
+        return self._state
+
+    @state.setter
+    def state(self, new_state: State) -> None:
+        self._state = new_state
+        self._state.engine = self
+        for k, v in self._state.kwargs.items():
+            setattr(self._state, k, v)
+            if k == "epoch" or k == "iteration":
+                delattr(self._state, k)
```

Comment:
This part of the code looks strange. Please, put a comment in the code and explain here, why we are doing that.

Reply:
Added a comment here anyway, I agree that I should clarify that.

Reply:
I think, we have to "repeat" the same steps as we do in the constructor:
```python
self._state = State(engine=self)
self.register_events(*Events, attr_to_events=State.attr_to_events)
```